xarray/core/dtypes.py

```diff
@@ -166,6 +166,65 @@ def is_datetime_like(dtype):
     return np.issubdtype(dtype, np.datetime64) or np.issubdtype(dtype, np.timedelta64)
 
 
+def isdtype(dtype, kind, xp=None):
```

Comment:
with numpy 2 we can make use of `numpy.isdtype`, which would replace `numpy_isdtype`. We still need to have a separate code path for the array api and extension array, because the array api does not support certain dtypes (like string, or datetime / timedelta).

(unit tests have been added since you wrote this, could you check if that's what you had in mind?)